<img align="right" style="padding-left:50px;" src="figures_wk4/data_cleaning.png" width=350><br>
### User Bias in Data Cleaning
For your homework assignment this week, we will explore how our treatment of our data can impact the quality of our results.

**Dataset:**
The data is a Salary Survey from AskAManager.org. It’s US-centric-ish but does allow for a range of country inputs.

A list of the corresponding survey questions can be found [here](https://www.askamanager.org/2021/04/how-much-money-do-you-make-4.html).

 

In [2]:
import pandas as pd

In [3]:
df= pd.read_csv('survey_data.csv')

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28108 entries, 0 to 28107
Data columns (total 18 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   timestamp  28108 non-null  object 
 1   q1         28108 non-null  object 
 2   q2         28033 non-null  object 
 3   q3         28107 non-null  object 
 4   q4         7273 non-null   object 
 5   q5         28108 non-null  object 
 6   q6         20793 non-null  float64
 7   q7         28108 non-null  object 
 8   q8         211 non-null    object 
 9   q9         3047 non-null   object 
 10  q10        28108 non-null  object 
 11  q11        23074 non-null  object 
 12  q12        28026 non-null  object 
 13  q13        28108 non-null  object 
 14  q14        28108 non-null  object 
 15  q15        27885 non-null  object 
 16  q16        27937 non-null  object 
 17  q17        27931 non-null  object 
dtypes: float64(1), object(17)
memory usage: 3.9+ MB


In [5]:
df.head()

timestamp     q1                             q2  \
0  4/27/2021 11:02:10  25-34   Education (Higher Education)   
1  4/27/2021 11:02:22  25-34              Computing or Tech   
2  4/27/2021 11:02:38  25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41  25-34                     Nonprofits   
4  4/27/2021 11:02:42  25-34  Accounting, Banking & Finance   

                                         q3   q4      q5      q6   q7   q8  \
0        Research and Instruction Librarian  NaN  55,000     0.0  USD  NaN   
1  Change & Internal Communications Manager  NaN  54,600  4000.0  GBP  NaN   
2                      Marketing Specialist  NaN  34,000     NaN  USD  NaN   
3                           Program Manager  NaN  62,000  3000.0  USD  NaN   
4                        Accounting Manager  NaN  60,000  7000.0  USD  NaN   

    q9             q10             q11          q12           q13  \
0  NaN   United States   Massachusetts       Boston     5-7 years   
1  NaN  United Kingdom             NaN    Cambridge  8 - 10 years   
2  NaN              US       Tennessee  Chattanooga   2 - 4 years   
3  NaN             USA       Wisconsin    Milwaukee  8 - 10 years   
4  NaN              US  South Carolina   Greenville  8 - 10 years   

           q14              q15         q16    q17  
0    5-7 years  Master's degree       Woman  White  
1    5-7 years   College degree  Non-binary  White  
2  2 - 4 years   College degree       Woman  White  
3    5-7 years   College degree       Woman  White  
4    5-7 years   College degree       Woman  White

### Assignment
Your goal for this assignment is to observe how your data treatment during the cleaning process can skew or bias the dataset.

Before diving right in, stop and read through the questions associated with the dataset. As you can see, they are either free-form text entries or categorical selections. Knowing this, perform some exploratory data analysis (EDA) to investigate the "state" of the dataset.

[Add as many code cell below here as needs]


<div class="alert alert-block alert-info">
Just putting a break into things so it's easier to figure out what is what and to draw the eye to it.

In [8]:
total = df.isnull().sum().sort_values(ascending=False)
total

q8           27897
q9           25061
q4           20835
q6            7315
q11           5034
q15            223
q17            177
q16            171
q12             82
q2              75
q3               1
q7               0
q5               0
q1               0
q10              0
q13              0
q14              0
timestamp        0
dtype: int64

In [9]:
df_marg=df.groupby(['q17', 'q16'])['q16'].count().reset_index(name='GenderCount')
df_marg=df_marg.sort_values(by='GenderCount', ascending=False).reset_index()
df_marg

index                                                q17         q16  \
0      106                                              White       Woman   
1      103                                              White         Man   
2        8                            Asian or Asian American       Woman   
3      104                                              White  Non-binary   
4       43                          Black or African American       Woman   
..     ...                                                ...         ...   
106     12  Asian or Asian American, Black or African Amer...  Non-binary   
107     81  Hispanic, Latino, or Spanish origin, White, An...  Non-binary   
108     53  Black or African American, Middle Eastern or N...       Woman   
109     83                 Middle Eastern or Northern African  Non-binary   
110     21  Asian or Asian American, Hispanic, Latino, or ...  Non-binary   

     GenderCount  
0          17931  
1           4473  
2           1021  
3            607  
4            562  
..           ...  
106            1  
107            1  
108            1  
109            1  
110            1  

[111 rows x 4 columns]

In [10]:
df_ed=df.groupby(['q15', 'q16'])['q16'].count().reset_index(name='EduCount')
df_ed=df_ed.sort_values(by='EduCount', ascending=False).reset_index()
df_ed

index                                 q15                            q16  \
0       3                      College degree                          Woman   
1      11                     Master's degree                          Woman   
2       0                      College degree                            Man   
3       8                     Master's degree                            Man   
4      24                        Some college                          Woman   
5      16                                 PhD                          Woman   
6      20  Professional degree (MD, JD, etc.)                          Woman   
7      21                        Some college                            Man   
8       1                      College degree                     Non-binary   
9       7                         High School                          Woman   
10     12                                 PhD                            Man   
11      4                         High School                            Man   
12      9                     Master's degree                     Non-binary   
13     17  Professional degree (MD, JD, etc.)                            Man   
14      2                      College degree  Other or prefer not to answer   
15     22                        Some college                     Non-binary   
16     10                     Master's degree  Other or prefer not to answer   
17     13                                 PhD                     Non-binary   
18     23                        Some college  Other or prefer not to answer   
19      5                         High School                     Non-binary   
20     18  Professional degree (MD, JD, etc.)                     Non-binary   
21     14                                 PhD  Other or prefer not to answer   
22     19  Professional degree (MD, JD, etc.)  Other or prefer not to answer   
23      6                         High School  Other or prefer not to answer   
24     15                                 PhD           Prefer not to answer   

    EduCount  
0      10233  
1       7123  
2       2729  
3       1410  
4       1353  
5       1097  
6       1092  
7        599  
8        372  
9        360  
10       277  
11       251  
12       214  
13       191  
14       154  
15        96  
16        81  
17        24  
18        23  
19        21  
20        18  
21        16  
22        12  
23         8  
24         1

In [11]:
df['q16'].value_counts()

q16
Woman                            21389
Man                               5502
Non-binary                         747
Other or prefer not to answer      298
Prefer not to answer                 1
Name: count, dtype: int64

In [12]:
df.q17.value_counts()

q17
White                                                                                                                                                                                                                           23235
Asian or Asian American                                                                                                                                                                                                          1410
Black or African American                                                                                                                                                                                                         694
Another option not listed here or prefer not to answer                                                                                                                                                                            625
Hispanic, Latino, or Spanish origin                                         

In [13]:
df.q1.value_counts()

q1
25-34         12668
35-44          9908
45-54          3193
18-24          1236
55-64           994
65 or over       95
under 18         14
Name: count, dtype: int64

<div class="alert alert-block alert-info">

**Question:** How would you describe the "state" of this dataset? Be specific and detailed in your answer. (Think paragraphs rather than sentences).

This data set is meant to represent the salary made by those who enter their data.  It is self reported.  <p>
The data is entered in either free form or by radio button.  In some instances the end user can select multiple radio buttons.
<p>
Initial impression of the 'state of the data':  the data is definitely dirty.  It appears that every single record has at least 1 null in it.  It is also evident that due to the free form entering, grouping the data will be difficult due to misspellings, differences in abbreviations (i.e. USA, US, U.S.A., U.S.), differences in capitalization (i.e. us, US), added information in the text fields (i.e. AUD (austrailian dollars), czek money).  With the exception of 2 features in the data, everything is an object, which will not lend itself easily to graphing or visual representation of the data.  Age (q1) has been pre grouped, and appears to have no nulls present.  While this field is a text field, it can be converted to a corresponding number for a better idea of the age groups that responded to the survey.  The data also has a field for how long the respondent has been working and one for how long they've been working in their field.  Like the age field, this will need to be converted to a numeric value before it will be useful in graphing.  
<p>
The gender (q16) feature demonstrates that there is an extreme imbalance between women and the rest of the genders.  This will create a skewed end result for any analysis or modeling that will be done with this data.  Moreover, the counts indicate that white women dominate this data set.  In reviewing the counts grouped with education level, it is evident that the women who responded to the survey also have a fairly high level of education.  
<p>
Prior to use, this data set will need the nulls cleaned up.  After the nulls are cleaned up, it might be helpful to clean up other text features (fix spelling and abbreviations).


<div class="alert alert-block alert-info">

Now, it is time to plan how you will clean up the dataset. You **are not** allowed to use any machine learning technique to clean the data. (No SMOTE! No machine learning! Or anything like that!)

**Question:** Based on your EDA above, detail how you would clean up this dataset. 
Things to consider: (This is not an exhaustive list)
- Are there columns that can't be effectively cleaned? If so, why?
- Are there columns that genuinely won't have a data value?
- Does it make sense to segment the dataset based on specific columns when determining how to handle the missing values?
- Are outliers a factor in this dataset?

Remember preserving as much of the data as possible is the goal. That means dropping rows with a missing value somewhere might not be the best idea.

[Add you answer to this markdown cell]

<b>  Based on the question page: </b>
<br>Q2 is the industry someone works in.  The nulls in this could be safely renamed as 'Other'  or 'Unknown', rather than deleting those 75 records.
<br>Q3 is a required field.  Not sure how a null snuck in.  This could likely be dropped without disrupting the data too much 
<br>Q4 is additional job title context.  This null in most records--only 8k of the records have a value here.  The data can be explored. however, it's not a required field and we can probably move forward without it.  
<br>Q6 is additional compensation.  There are 7k null values.  This can be imputed to 0.  It may not be 100% accurate, yet, it is safer than using the mean/median of the other records.
<br>Q8 is currency not listed.  This should be looked at for distinct values.  It is likely that we can replace 'Other' in Q7 with this value.  If Q8 has too many values or has values that are nonsensical, we should be able to drop it without too much consequence.  
<br>Q9 This is additional info about income. There's only approx 3k records with data.  Review what is there and see if it can be combined with another field.  If not, pretty sure it can be dropped.  
<br>Q11 This is the state that the respondent works in.  Not a required field and many foreign countries don't have states or provinces.  I'll need to verify that the Nulls are indeed for foregin locations.  If they aren't, I should be able to fill in the null based on Q12.  If they are, I should be able to safely set this to 'Unknown', since there are only 5k that are null. 
<br>Q12 This is the city where the respondant works.  It is a required field, yet we still have 82 of them.  I'll need to look at the records themselves to see if those can be dropped. 
<br>Q15 Highest level of education.  I can replace nulls with 'unknown'.  
<br>Q16 Gender.  This can be set to 'Other or prefer not to answer'
<br>Q17 Race.  This can be set to 'Another option not listed here or prefer not to answer'.
<p>

Possible analysis directions:
<br>***may need to convert wages to USD to accommodate for different currency type
<br>Wages by industry
<br>Wages by job title
<br>Wages by Country/State
<br>Wages by city
<br>Wages Education
<br>Wages by Gender
<br>Wages by Race
<br>Wages by Age group 
<p>
Other EDA
<br>Counts by Gender, by Race, age group.

<div class="alert alert-block alert-info">

Based on the plan the you described above, go ahead and clean up the dataset.

[Add as many code cell below here as needs]

Bring in the packages 

In [20]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from ydata_profiling import ProfileReport

address q2 and set the values to Other: Unknown

In [22]:
df['q2'].value_counts()

q2
Computing or Tech                          4711
Education (Higher Education)               2466
Nonprofits                                 2420
Health care                                1899
Government and Public Administration       1893
                                           ... 
Gaming (Gambling)                             1
Regulatory Affairs- nutraceuticals            1
Manufacturing : corporate admin support       1
Real Estate Investment Support                1
Social networks                               1
Name: count, Length: 1220, dtype: int64

In [23]:
df['q2'] = df['q2'].fillna('Other: Unknown')

review the record that has q3 null to confirm it can be dropped

In [25]:
df[df['q3'].isnull()]

timestamp     q1  \
27996  11/29/2023 19:59:34  18-24   

                                                 q2   q3  \
27996  I'm currently a student and don't have a job  NaN   

                               q4 q5   q6   q7   q8  \
27996  I don't have one right now  0  0.0  USD  NaN   

                                 q9    q10           q11  q12          q13  \
27996  There is no income right now  U.S.A  Pennsylvania  NaN  2 - 4 years   

               q14           q15  q16                        q17  
27996  2 - 4 years  Some college  Man  Black or African American

This record doesn't add enough of the dataset to keep it.  We can drop the record safely.

In [27]:
df.dropna(subset = ['q3'])

timestamp     q1                             q2  \
0       4/27/2021 11:02:10  25-34   Education (Higher Education)   
1       4/27/2021 11:02:22  25-34              Computing or Tech   
2       4/27/2021 11:02:38  25-34  Accounting, Banking & Finance   
3       4/27/2021 11:02:41  25-34                     Nonprofits   
4       4/27/2021 11:02:42  25-34  Accounting, Banking & Finance   
...                    ...    ...                            ...   
28103  12/29/2024 23:50:26  25-34                  Entertainment   
28104    1/3/2025 16:03:10  18-24   Engineering or Manufacturing   
28105   1/16/2025 10:00:47  35-44                  Entertainment   
28106   1/23/2025 19:38:04  35-44                     Nonprofits   
28107   1/27/2025 21:02:19  45-54              Computing or Tech   

                                                   q3   q4       q5       q6  \
0                  Research and Instruction Librarian  NaN   55,000      0.0   
1            Change & Internal Communications Manager  NaN   54,600   4000.0   
2                                Marketing Specialist  NaN   34,000      NaN   
3                                     Program Manager  NaN   62,000   3000.0   
4                                  Accounting Manager  NaN   60,000   7000.0   
...                                               ...  ...      ...      ...   
28103                                   Junior Editor  NaN    72800      NaN   
28104                           Applications Engineer  NaN   100000  21000.0   
28105  Product communication and education specialist  NaN  1000000      NaN   
28106                      Director of content design  NaN   132000      NaN   
28107                                             CMO  NaN   279000      0.0   

          q7   q8   q9              q10             q11          q12  \
0        USD  NaN  NaN    United States   Massachusetts       Boston   
1        GBP  NaN  NaN   United Kingdom             NaN    Cambridge   
2        USD  NaN  NaN               US       Tennessee  Chattanooga   
3        USD  NaN  NaN              USA       Wisconsin    Milwaukee   
4        USD  NaN  NaN               US  South Carolina   Greenville   
...      ...  ...  ...              ...             ...          ...   
28103    USD  NaN  NaN    United States      California  Los Angeles   
28104    USD  NaN  NaN              USA           Texas       Dallas   
28105  Other  Czk  NaN  Česká republika             NaN       Prague   
28106    USD  NaN  NaN              USA        Missouri  Kansas City   
28107    USD  NaN  NaN              USA      Washington      seattle   

                  q13             q14              q15         q16  \
0           5-7 years       5-7 years  Master's degree       Woman   
1        8 - 10 years       5-7 years   College degree  Non-binary   
2         2 - 4 years     2 - 4 years   College degree       Woman   
3        8 - 10 years       5-7 years   College degree       Woman   
4        8 - 10 years       5-7 years   College degree       Woman   
...               ...             ...              ...         ...   
28103       5-7 years       5-7 years  Master's degree       Woman   
28104  1 year or less  1 year or less  Master's degree       Woman   
28105   21 - 30 years     2 - 4 years      High School       Woman   
28106   11 - 20 years   11 - 20 years  Master's degree       Woman   
28107   21 - 30 years   11 - 20 years  Master's degree       Woman   

                                              q17  
0                                           White  
1                                           White  
2                                           White  
3                                           White  
4                                           White  
...                                           ...  
28103  Hispanic, Latino, or Spanish origin, White  
28104                                       White  
28105                                       White  
28106        

check q4 to see what it actually adds to the dataset and if we can reasonably drop it

In [29]:
df[df['q4'].notnull()]

timestamp     q1                                    q2  \
7       4/27/2021 11:03:00  25-34         Education (Primary/Secondary)   
8       4/27/2021 11:03:01  45-54                     Computing or Tech   
13      4/27/2021 11:03:10  25-34                                   Law   
26      4/27/2021 11:03:38  25-34                            Nonprofits   
35      4/27/2021 11:03:55  35-44          Education (Higher Education)   
...                    ...    ...                                   ...   
28093   11/18/2024 7:24:09  45-54                     Computing or Tech   
28094  11/18/2024 11:08:58  18-24                     Computing or Tech   
28095  11/22/2024 15:35:09  25-34                  Hospitality & Events   
28097   12/5/2024 17:12:41  25-34         Education (Primary/Secondary)   
28099   12/6/2024 19:52:27  35-44  Government and Public Administration   

                              q3  \
7                      Librarian   
8                Systems Analyst   
13      Legal Aid Staff Attorney   
26              Program Director   
35            Senior IRB Analyst   
...                          ...   
28093             CYBER SECURITY   
28094  Cyber Security Management   
28095            Revenue Manager   
28097     Elementary Art Teacher   
28099          Station Commander   

                                                      q4       q5       q6  \
7                                        High school, FT   50,000      NaN   
8                           Data developer/ETL Developer  112,000  10000.0   
13                                   non-profit law firm   52,000      0.0   
26     I manage staff members who each manage 8-15 ot...   62,000      NaN   
35     State university, higher ed research adminstra...   68,000      0.0   
...                                                  ...      ...      ...   
28093                                                 NO       40   1200.0   
28094                                      It sicherheit  1500000  50000.0   
28095  Hotel with between 500-600 rooms and 75,000+ s...    65000   1200.0   
28097                              Public charter school    60000      NaN   
28099       Station Commander of an Army Recruiting Team    70000  40000.0   

        q7     q8                                                 q9  \
7      USD    NaN                                                NaN   
8      USD    NaN                                                NaN   
13     USD    NaN                                                NaN   
26     USD    NaN                                                NaN   
35     USD    NaN                                                NaN   
...    ...    ...                                                ...   
28093  EUR  other                                                 no   
28094  EUR    NaN                                                NaN   
28095  USD    NaN                                                NaN   
28097  USD    NaN                                                NaN   
28099  USD    NaN  Housing Allowance, Food Allowance, and Special...   

                 q10                   q11                             q12  \
7      United States               Arizona                            Yuma   
8                 US              Missouri                       St. Louis   
13     United States             Minnesota                      Saint Paul   
26                US  District of Columbia                   Washington DC   
35               USA        North Carolina               Research Triangle   
...              ...                   ...                             ...   
28093        germany   Delaware, Louisiana                      Düsseldorf   
28094        Germany               Indiana                         germany   
28095             US             Tennessee                       Nashville   
28097            USA             Tennessee                       Nashville   
28099  United States   

In [30]:
#it looks like it is safe to drop the column.  While the data is interesting, it won't add a lot to analysis
df.drop('q4', axis=1, inplace=True)

q5 holds the salary data.  q6 is additional compensation.  Once we've dealt q6, we can deal with q5

In [32]:
df[df['q6'].notnull()]

timestamp     q1                                    q2  \
0       4/27/2021 11:02:10  25-34          Education (Higher Education)   
1       4/27/2021 11:02:22  25-34                     Computing or Tech   
3       4/27/2021 11:02:41  25-34                            Nonprofits   
4       4/27/2021 11:02:42  25-34         Accounting, Banking & Finance   
6       4/27/2021 11:02:51  25-34                            Publishing   
...                    ...    ...                                   ...   
28099   12/6/2024 19:52:27  35-44  Government and Public Administration   
28100  12/11/2024 14:53:19  25-34                           Health care   
28102   12/20/2024 6:16:27  55-64  Government and Public Administration   
28104    1/3/2025 16:03:10  18-24          Engineering or Manufacturing   
28107   1/27/2025 21:02:19  45-54                     Computing or Tech   

                                             q3        q5         q6     q7  \
0            Research and Instruction Librarian    55,000        0.0    USD   
1      Change & Internal Communications Manager    54,600     4000.0    GBP   
3                               Program Manager    62,000     3000.0    USD   
4                            Accounting Manager    60,000     7000.0    USD   
6                          Publishing Assistant    33,000     2000.0    USD   
...                                         ...       ...        ...    ...   
28099                         Station Commander     70000    40000.0    USD   
28100                                Pharmacist  24000000  2000000.0  Other   
28102                      HMRC Engagement Lead     25000        0.0    GBP   
28104                     Applications Engineer    100000    21000.0    USD   
28107                                       CMO    279000        0.0    USD   

        q8                                                 q9             q10  \
0      NaN                                                NaN   United States   
1      NaN                                                NaN  United Kingdom   
3      NaN                                                NaN             USA   
4      NaN                                                NaN              US   
6      NaN                                                NaN             USA   
...    ...                                                ...             ...   
28099  NaN  Housing Allowance, Food Allowance, and Special...   United States   
28100  TZS            full time pharmacist and superintendent        Tanzania   
28102  NaN                                                NaN  United Kingdom   
28104  NaN                                                NaN             USA   
28107  NaN                                                NaN             USA   

                    q11                             q12             q13  \
0         Massachusetts                          Boston       5-7 years   
1                   NaN                       Cambridge    8 - 10 years   
3             Wisconsin                       Milwaukee    8 - 10 years   
4        South Carolina                      Greenville    8 - 10 years   
6        South Carolina                        Columbia     2 - 4 years   
...                 ...                             ...             ...   
28099  Florida, Georgia  Tifton, Valdosta, Jacksonville   21 - 30 years   
28100               NaN                   Dar es Salaam     2 - 4 years   
28102               NaN                         Suffolk   21 - 30 years   
28104             Texas                          Dallas  1 year or less   
28107        Washington                         seattle   21 - 30 years   

                  q14              q15         q16  \
0           5-7 years  Master's degree       Woman   
1           5-7 years   College degree  Non-binary   
3           5-7 years   College degree       Woman   
4           5-7 years   College degree       Woman   
6         2 - 4 years  

replace the nulls with 0, convert q5 to float and then create a new column that is q5+q6

In [34]:
df['q6'] = df['q6'].fillna(0)

In [35]:
df['q5'] = df['q5'].apply(lambda x : x.replace(',','')).astype(float)

In [36]:
df['Totalq5']=df['q5']+df['q6']

q8 reflects the monetary denomination.  This is a free form text field that appears to be all over in terms of what has been entered.  Cleaning that up will allow for easier grouping and segregation of compensation types.

In [38]:
df[df['q8'].notnull()]

timestamp     q1                                         q2  \
434     4/27/2021 11:09:54  25-34                          Computing or Tech   
603     4/27/2021 11:12:24  45-54                                Health care   
752     4/27/2021 11:14:29  45-54  Public Library (technically City Govt.?)    
766     4/27/2021 11:14:39  25-34                          Recruitment or HR   
776     4/27/2021 11:14:45  45-54               Education (Higher Education)   
...                    ...    ...                                        ...   
28077   9/19/2024 19:16:28  25-34                                Social Work   
28093   11/18/2024 7:24:09  45-54                          Computing or Tech   
28096   11/24/2024 8:08:37  25-34                     Business or Consulting   
28100  12/11/2024 14:53:19  25-34                                Health care   
28105   1/16/2025 10:00:47  35-44                              Entertainment   

                                                   q3          q5         q6  \
434                                      Data analyst    885000.0        0.0   
603                                        Exec admin   1080000.0   223000.0   
752                          Administrative Librarian     76302.0        0.0   
766                Lead Payroll & Benefits Specialist     64000.0     6400.0   
776                          Sr Technical Support Rep     53000.0        0.0   
...                                               ...         ...        ...   
28077                                      Counsellor     40000.0        0.0   
28093                                  CYBER SECURITY        40.0     1200.0   
28096                                  BI Consultant         10.0        0.0   
28100                                      Pharmacist  24000000.0  2000000.0   
28105  Product communication and education specialist   1000000.0        0.0   

          q7                                                 q8  \
434    Other                                                INR   
603    Other                                     Peso Argentino   
752      USD                                         $76,302.34   
766      USD  My bonus is based on performance up to 10% of ...   
776      USD  I work for an online state university, managin...   
...      ...                                                ...   
28077  Other                                                SGD   
28093    EUR                                              other   
28096  Other                                            Rupees    
28100  Other                                                TZS   
28105  Other                                                Czk   

                                                      q9              q10  \
434                                                  NaN            India   
603                           Exchange about 1USD=165ARS        Argentina   
752                                                  NaN              USA   
766                                                  NaN               US   
776    I have excellent health benefits, time off, si...              USA   
...                                                  ...              ...   
28077                                                NaN        Singapore   
28093                                                 no          germany   
28096                                                NaN            India   
28100            full time pharmacist and superintendent         Tanzania   
28105                                                NaN  Česká republika   

                       q11                   q12            q13  \
434                    NaN                  Pune    2 - 4 years   
603                    NaN  City of Buenos Aires  11 - 20 years   
752                Arizona             Glendale   21 - 30 years   
766               New York           Schenectady      5-7 years   
776             New Jersey    

In [39]:
df[ (df['q7'] =='Other') & (df['q8'].isnull()) ]

timestamp     q1                                    q2  \
9344   4/27/2021 17:44:22  45-54  Government and Public Administration   
18843   4/29/2021 3:50:24  18-24                           Health care   
18882   4/29/2021 4:30:05  25-34                     Computing or Tech   
18904   4/29/2021 4:59:14  45-54                               Mining    
27959  4/29/2023 15:24:07  18-24          Engineering or Manufacturing   

                            q3       q5   q6     q7   q8   q9            q10  \
9344   Contracts Administrator  76000.0  0.0  Other  NaN  NaN  United States   
18843   Microbiologist officer  48000.0  0.0  Other  NaN   RM       Malaysia   
18882             Data Analyst  97000.0  0.0  Other  NaN  NaN       Malaysia   
18904          Office manager   55140.0  0.0  Other  NaN  NaN           USA    
27959            Data Engineer  10000.0  0.0  Other  NaN  NaN          India   

              q11           q12             q13             q14  \
9344   California  Berkeley, CA   11 - 20 years       5-7 years   
18843         NaN         Tawau  1 year or less  1 year or less   
18882         NaN    Georgetown     2 - 4 years     2 - 4 years   
18904      Nevada          Elko   11 - 20 years   11 - 20 years   
27959         NaN      Banglore  1 year or less  1 year or less   

                  q15         q16  \
9344     Some college       Woman   
18843    Some college         Man   
18882  College degree  Non-binary   
18904    Some college       Woman   
27959             NaN         Man   

                                                     q17  Totalq5  
9344                                               White  76000.0  
18843                            Asian or Asian American  48000.0  
18882  White, Another option not listed here or prefe...  97000.0  
18904                                              White  55140.0  
27959                            Asian or Asian American  10000.0

In [40]:
df.loc[9344, 'q7'] = 'USD'
df.loc[18904, 'q7'] = 'USD'
df.loc[27959, 'q7'] = 'RUP'
df.loc[18843, 'q7'] = 'RM'
df.loc[18882, 'q7'] = 'RM'

In [41]:
df.loc[df["q8"] == "Indian rupees", "q7"] = "RUP"
df.loc[df["q8"] == "Rupees", "q7"] = "RUP"
df.loc[df["q8"] == "US Dollar", "q7"] = "USD"
df.loc[df["q8"] == "American Dollars", "q7"] = "USD"
df.loc[df["q8"] == "NIS (new Israeli shekel)", "q7"] = "NIS"
df.loc[df["q8"] == "PLN (Polish zloty)", "q7"] = "PLN"

In [42]:
df.loc[df["q8"] == "Norwegian kroner (NOK)", "q7"] = "NOK"
df.loc[df["q8"] == "RMB (chinese yuan)", "q7"] = "RMB"
df.loc[df["q8"] == "KRW (Korean Won)", "q7"] = "KRW"
df.loc[df["q8"] == "ILS (Shekel)", "q7"] = "ILS"
df.loc[df["q8"] == "China RMB", "q7"] = "RMB"
df.loc[df["q8"] == "AUD Australian", "q7"] = "AUD"
df.loc[df["q8"] == "PLN (Zwoty)", "q7"] = "PLN"
df.loc[df["q8"] == "czech crowns", "q7"] = "CZK"
df.loc[df["q8"] == "Mexican pesos", "q7"] = "MXN"

In [43]:
df.loc[df["q8"] == "Peso Argentino", "q7"] = "ARS"
df.loc[df["q8"] == "Philippine Peso", "q7"] = "PHP"
df.loc[df["q8"] == "Taiwanese dollars", "q7"] = "TWD"
df.loc[df["q8"] == "THAI BAHT", "q7"] = "THB"
df.loc[df["q8"] == "AUD Australian", "q7"] = "AUD"
df.loc[df["q8"] == "Polish Złoty", "q7"] = "PLN"
df.loc[df["q10"] == "Australia", "q8"] = "AUD"
df.loc[df["q8"] == "Philippine peso (PHP)", "q7"] = "PHP"

In [44]:
df['q7'] = df.apply(lambda x: x['q8'] if str(x['q8'])in ['THB','Bdt','ZAR','RUP','ARS','Php','PLN','CNY','ILS/NIS','AUD','BRL','IDR','Sgd','NZD','DKK'] else x['q7'], axis=1)

In [45]:
df['q7'] = df.apply(lambda x: x['q8'] if str(x['q8'])in ['MXN','NIS','ZAR','RMB','TWD','PHP','KRW','IDR','Nok','ILS','DKK','IDR','LKR','NZD','SAR'] else x['q7'], axis=1)

In [46]:
df['q7'] = df.apply(lambda x: x['q8'] if str(x['q8'])in ['USD', 'MYR', 'CHF','NOK','INR','SGD','USD','SEK','EUR','BR$','TRY','DKK','Dkk','NGN','COP','TTD','TZS','Czk','CZK','GBP','PLN','RUP','BRL (R$)'] else x['q7'], axis=1)

In [47]:
df.loc[11454, 'q7'] = 'IDR'
df.loc[11737, 'q7'] = 'DKK'
df.loc[13218, 'q7'] = 'AUD'
df.loc[14915, 'q7'] = 'USD'
df.loc[16395, 'q7'] = 'ARS'
df.loc[16627, 'q7'] = 'ILS'
df.loc[17113, 'q7'] = 'PHP'
df.loc[18930, 'q7'] = 'RM'

In [48]:
df.loc[19250, 'q7'] = 'ARS'
df.loc[19323, 'q7'] = 'PHP'
df.loc[21322, 'q7'] = 'INR'
df.loc[21350, 'q7'] = 'NTD'
df.loc[21545, 'q7'] = 'DKK'
df.loc[22826, 'q7'] = 'CAD'
df.loc[22871, 'q7'] = 'KRW'
df.loc[23123, 'q7'] = 'DKK'
df.loc[24198, 'q7'] = 'EUR'
df.loc[24843, 'q7'] = 'SGD'

In [49]:
df.loc[24884, 'q7'] = 'NTD'
df.loc[24949, 'q7'] = 'MXN'
df.loc[25548, 'q7'] = 'THB'
df.loc[26424, 'q7'] = 'THB'
df.loc[28096, 'q7'] = 'RUP'
df.loc[27834, 'q7'] = 'SGD'
df.loc[27191, 'q7'] = 'HRK'
df.loc[20373, 'q7'] = 'ILS'

In [50]:
df[ (df['q7'] =='Other') ]

Empty DataFrame
Columns: [timestamp, q1, q2, q3, q5, q6, q7, q8, q9, q10, q11, q12, q13, q14, q15, q16, q17, Totalq5]
Index: []

Now that Other has been updated, we can drop q8

In [52]:
df.drop('q8', axis=1, inplace=True)

In [53]:
df[df['q9'].notnull()]

timestamp          q1                                    q2  \
9       4/27/2021 11:03:02       35-44         Accounting, Banking & Finance   
32      4/27/2021 11:03:45       25-34         Accounting, Banking & Finance   
44      4/27/2021 11:04:04       25-34                     Computing or Tech   
62      4/27/2021 11:04:29  65 or over              Property or Construction   
65      4/27/2021 11:04:33       25-34          Education (Higher Education)   
...                    ...         ...                                   ...   
28069   9/10/2024 23:58:45       35-44                     Computing or Tech   
28078   9/23/2024 17:40:20       18-24              Leisure, Sport & Tourism   
28093   11/18/2024 7:24:09       45-54                     Computing or Tech   
28099   12/6/2024 19:52:27       35-44  Government and Public Administration   
28100  12/11/2024 14:53:19       25-34                           Health care   

                             q3          q5         q6   q7  \
9             Senior Accountant     45000.0        0.0  USD   
32            Financial Advisor    230000.0        0.0  USD   
44         Intelligence Analyst    110000.0    20000.0  USD   
62              Safety Director    104000.0    25000.0  USD   
65          Research Programmer     70000.0        0.0  USD   
...                         ...         ...        ...  ...   
28069  Business Systems Analyst     90000.0     9000.0  USD   
28078                  dockhand     33280.0    10500.0  USD   
28093            CYBER SECURITY        40.0     1200.0  EUR   
28099         Station Commander     70000.0    40000.0  USD   
28100                Pharmacist  24000000.0  2000000.0  TZS   

                                                      q9            q10  \
9                            I work for a Charter School  United States   
32                                      Commission based            USA   
44                         Around 20,000 a year in stock            USA   
62                       Extra money comes from overtime            USA   
65     Technically 37.5 hours, but ends up being pret...            USA   
...                                                  ...            ...   
28069              10% of the salary if te goals are met            USA   
28078                                     I recieve tips  United States   
28093                                                 no        germany   
28099  Housing Allowance, Food Allowance, and Special...  United States   
28100            full time pharmacist and superintendent       Tanzania   

                       q11                             q12               q13  \
9                  Florida                      Palm Coast     21 - 30 years   
32                Missouri                       St. Louis     11 - 20 years   
44                Virginia                   Arlington, VA      8 - 10 years   
62                 Georgia                         Atlanta  41 years or more   
65            Rhode Island                      Providence         5-7 years   
...                    ...                             ...               ...   
28069               Nevada                       Las Vegas      8 - 10 years   
28078              Georgia                      Greensboro    1 year or less   
28093  Delaware, Louisiana                      Düsseldorf       2 - 4 years   
28099     Florida, Georgia  Tifton, Valdosta, Jacksonville     21 - 30 years   
28100                  NaN                   Dar es Salaam       2 - 4 years   

                  q14              q15         q16  \
9       21 - 30 years   College degree       Woman   
32      11 - 20 years   College degree       Woman   
44       8 - 10 years  Master's degree         Man   
62      21 - 30 years   College degree         Man   
65        2 - 4 years  Master's degree  Non-binary   
...               ...              ...         ...   
28069     2 - 4 years  Master's degree       Woman   
28078

q9 is additional info about the respondents compensation is for.  This doesn't add a lot to the data set.  It should be safe to drop it

In [55]:
df.drop('q9', axis=1, inplace=True)

In [56]:
df[df['q11'].isnull()]

timestamp     q1                                    q2  \
1       4/27/2021 11:02:22  25-34                     Computing or Tech   
10      4/27/2021 11:03:03  25-34                            Nonprofits   
14      4/27/2021 11:03:11  18-24                           Health care   
15      4/27/2021 11:03:11  35-44        Utilities & Telecommunications   
21      4/27/2021 11:03:24  35-44                            Nonprofits   
...                    ...    ...                                   ...   
28096   11/24/2024 8:08:37  25-34                Business or Consulting   
28098    12/6/2024 5:45:00  18-24                        Other: Unknown   
28100  12/11/2024 14:53:19  25-34                           Health care   
28102   12/20/2024 6:16:27  55-64  Government and Public Administration   
28105   1/16/2025 10:00:47  35-44                         Entertainment   

                                                   q3          q5         q6  \
1            Change & Internal Communications Manager     54600.0     4000.0   
10                                     Office Manager     47500.0        0.0   
14                           Patient care coordinator     32000.0        0.0   
15                  Quality And Compliance Specialist     24000.0      500.0   
21            Program Coordinator & Assistant Editor      50000.0        0.0   
...                                               ...         ...        ...   
28096                                  BI Consultant         10.0        0.0   
28098                                   data engineer         0.0        0.0   
28100                                      Pharmacist  24000000.0  2000000.0   
28102                            HMRC Engagement Lead     25000.0        0.0   
28105  Product communication and education specialist   1000000.0        0.0   

        q7              q10  q11            q12             q13  \
1      GBP   United Kingdom  NaN      Cambridge    8 - 10 years   
10     USD    United States  NaN     Boston, MA       5-7 years   
14     CAD           Canada  NaN         Remote  1 year or less   
15     GBP  United Kingdom   NaN        Lincoln   11 - 20 years   
21     USD    United States  NaN        Atlanta       5-7 years   
...    ...              ...  ...            ...             ...   
28096  RUP            India  NaN          Noida     2 - 4 years   
28098  USD          Morocco  NaN          Rabat  1 year or less   
28100  TZS         Tanzania  NaN  Dar es Salaam     2 - 4 years   
28102  GBP   United Kingdom  NaN        Suffolk   21 - 30 years   
28105  Czk  Česká republika  NaN         Prague   21 - 30 years   

                  q14              q15         q16  \
1           5-7 years   College degree  Non-binary   
10          5-7 years   College degree       Woman   
14     1 year or less   College degree       Woman   
15          5-7 years   College degree         Man   
21        2 - 4 years              PhD       Woman   
...               ...              ...         ...   
28096     2 - 4 years   College degree         Man   
28098  1 year or less  Master's degree         Man   
28100     2 - 4 years   College degree       Woman   
28102     2 - 4 years     Some college       Woman   
28105     2 - 4 years      High School       Woman   

                                                     q17     Totalq5  
1                                                  White     58600.0  
10                                                 White     47500.0  
14                                                 White     32000.0  
15                                                 White     24500.0  
21                                                 White     50000.0  
...                                                  ...         ...  
28096  Another option not listed here or prefer not t...        10.0  
28098                                              White         0.0  
28100                          Black or African American  2600

q11 is the state the respondent works in, if applicable.  In some cases, the state is present in q12, however, it doesn't make sense to go through 5034 records to see which ones have the state in the city feature.  We can set it to unknown.  Same is true for q12.

In [58]:
df['q11'] = df['q11'].fillna('Unknown')

In [59]:
df[df['q12'].isnull()]

timestamp        q1  \
148     4/27/2021 11:05:54     25-34   
700     4/27/2021 11:13:46     25-34   
904     4/27/2021 11:16:49     45-54   
944     4/27/2021 11:17:32     35-44   
1396    4/27/2021 11:24:58     35-44   
...                    ...       ...   
27248  10/26/2021 13:43:53     18-24   
27901   11/15/2022 4:28:09     55-64   
27972   8/21/2023 22:47:22  under 18   
27996  11/29/2023 19:59:34     18-24   
28020   3/19/2024 17:40:39     25-34   

                                                 q2  \
148                    Education (Higher Education)   
700                                       Libraries   
904                               Computing or Tech   
944            Government and Public Administration   
1396           Government and Public Administration   
...                                             ...   
27248                                      Science    
27901                                   Health care   
27972                 Accounting, Banking & Finance   
27996  I'm currently a student and don't have a job   
28020                                      Student    

                                         q3        q5       q6   q7  \
148    English as a Second Language Teacher   18000.0      0.0  EUR   
700      Supervisor of Children's Services    60800.0      0.0  USD   
904        senior primary systems engineer   117000.0      6.0  USD   
944                      Disability support   57500.0      0.0  CAD   
1396          Health science administrator   116000.0      0.0  USD   
...                                     ...       ...      ...  ...   
27248                Laboratory technician    22000.0      0.0  GBP   
27901        Director, clinical programming  101400.0  20000.0  GBP   
27972                 McDonalds Crew Member   40000.0      0.0  USD   
27996                                   NaN       0.0      0.0  USD   
28020                              Student        0.0      0.0  USD   

                  q10           q11  q12             q13             q14  \
148             Spain       Unknown  NaN    8 - 10 years       5-7 years   
700     United States   Connecticut  NaN    8 - 10 years       5-7 years   
904               usa    California  NaN   31 - 40 years   11 - 20 years   
944            Canada       Unknown  NaN       5-7 years     2 - 4 years   
1396              Usa      Maryland  NaN   11 - 20 years     2 - 4 years   
...               ...           ...  ...             ...             ...   
27248  United kingdom       Unknown  NaN     2 - 4 years  1 year or less   
27901              Uk       Unknown  NaN   31 - 40 years   31 - 40 years   
27972             USA     Minnesota  NaN  1 year or less  1 year or less   
27996           U.S.A  Pennsylvania  NaN     2 - 4 years     2 - 4 years   
28020              US      Illinois  NaN  1 year or less  1 year or less   

                   q15                            q16  \
148     College degree                          Woman   
700    Master's degree                          Woman   
904     College degree                          Woman   
944     College degree                          Woman   
1396               PhD                          Woman   
...                ...                            ...   
27248  Master's degree                          Woman   
27901  Master's degree                          Woman   
27972      High School                          Woman   
27996     Some college                            Man   
28020   College degree  Other or prefer not to answer   

                             q17   Totalq5  
148                        White   18000.0  
700                        White   60800.0  
904    Black or African American  117006.0  
944                        White   57500.0  
1396                       White  116000.0  
...                          ...       ...  
27248                      White   22000.0  
27901                      White  121400.0  
27972    Asian or A

In [60]:
df['q12'] = df['q12'].fillna('Unknown')

In [61]:
df[df['q15'].isnull()]

timestamp     q1                              q2  \
13     4/27/2021 11:03:10  25-34                             Law   
24     4/27/2021 11:03:31  25-34                  Public Library   
192    4/27/2021 11:06:27  35-44    Education (Higher Education)   
241    4/27/2021 11:07:12  35-44                           Sales   
659    4/27/2021 11:13:14  45-54  Utilities & Telecommunications   
...                   ...    ...                             ...   
28035  4/23/2024 23:39:34  18-24    Education (Higher Education)   
28043   5/27/2024 2:47:26  25-34               Computing or Tech   
28049  6/28/2024 14:00:00  25-34    Education (Higher Education)   
28053  7/12/2024 18:23:10  18-24                      Nonprofits   
28089  11/11/2024 7:28:08  18-24     Marketing, Advertising & PR   

                               q3         q5      q6   q7             q10  \
13       Legal Aid Staff Attorney    52000.0     0.0  USD   United States   
24                Teen Librarian     44500.0     0.0  USD              US   
192           Associate Director     63000.0  2000.0  USD  United States    
241                 Sales support    23050.0     0.0  GBP         England   
659             Financial Analyst    87360.0     0.0  USD             USA   
...                           ...        ...     ...  ...             ...   
28035                     Teacher    13860.0     0.0  USD           Chile   
28043                        Lead  2600000.0     0.0  USD             USA   
28049                        fasd        5.0     0.0  EUR              ff   
28053    Adminstration Specialist     1641.0     0.0  CAD          Taiwan   
28089  email marketing specialist     4740.0    11.0  USD           india   

                                                     q11         q12  \
13                                             Minnesota  Saint Paul   
24                                               Florida   Bradenton   
192                                                Texas     El Paso   
241                                              Unknown      London   
659                                                Texas     Houston   
...                                                  ...         ...   
28035                                            Unknown        Stgo   
28043                                            Florida     Newyork   
28049  Colorado, Delaware, New Jersey, West Virginia,...          ff   
28053                                            Unknown      Taipei   
28089                                            Unknown     chennai   

                  q13             q14  q15    q16  \
13        2 - 4 years     2 - 4 years  NaN  Woman   
24          5-7 years     2 - 4 years  NaN  Woman   
192     11 - 20 years    8 - 10 years  NaN  Woman   
241     11 - 20 years    8 - 10 years  NaN  Woman   
659     21 - 30 years   11 - 20 years  NaN  Woman   
...               ...             ...  ...    ...   
28035  1 year or less  1 year or less  NaN    NaN   
28043    8 - 10 years     2 - 4 years  NaN    Man   
28049   21 - 30 years       5-7 years  NaN    NaN   
28053  1 year or less  1 year or less  NaN    Man   
28089     2 - 4 years     2 - 4 years  NaN    Man   

                                       q17    Totalq5  
13                                   White    52000.0  
24                                   White    44500.0  
192    Hispanic, Latino, or Spanish origin    65000.0  
241                                  White    23050.0  
659              Black or African American    87360.0  
...                                    ...        ...  
28035                                  NaN    13860.0  
28043                                  NaN  2600000.0  
28049                                  NaN        5.0  
28053              Asian or Asian American     1641.0  
28089              Asian or Asian American     4751.0  

[223 rows x 16 columns]

There's a relatively low number of records that have no value for education.  These can be set to Unknown.
<br>
For q16 (gender), we can safely set the nulls to Other or Prefer not to answer
<br> For q17 (race) we can safely set the nulls to Another option not listed here or prefer not to answer

In [63]:
df['q15'] = df['q15'].fillna('Unknown')

In [64]:
df['q16'] = df['q16'].fillna('Other or prefer not to answer')

In [65]:
df['q17'] = df['q17'].fillna('Another option not listed here or prefer not to answer')

Okay, nulls have been taken care of.  


In [67]:
df.q10.value_counts()

q10
United States      9004
USA                7946
US                 2612
Canada             1572
United States       668
                   ... 
IS                    1
United Kingdomk       1
 New Zealand          1
Cuba                  1
Česká republika       1
Name: count, Length: 382, dtype: int64

In [68]:
df.loc[df["q10"] == "United States", "q10"] = "USA"

In [69]:
df.loc[df["q10"] == "US", "q10"] = "USA"

In [70]:
df.loc[df["q10"] == "United States ", "q10"] = "USA"
df.loc[df["q10"] == "United states", "q10"] = "USA"
df.loc[df["q10"] == "United statew", "q10"] = "USA"
df.loc[df["q10"] == "For the United States government, but posted overseas", "q10"] = "USA"
df.loc[df["q10"] == "U.S.", "q10"] = "USA"
df.loc[df["q10"] == "USA ", "q10"] = "USA"
df.loc[df["q10"] == "Usa", "q10"] = "USA"
df.loc[df["q10"] == "usa", "q10"] = "USA"
df.loc[df["q10"] == "United States of America", "q10"] = "USA"
df.loc[df["q10"] == "United Kingdom", "q10"] = "UK"
df.loc[df["q10"] == "Englang", "q10"] = "UK"
df.loc[df["q10"] == "Wales (United Kingdom)", "q10"] = "UK"
df.loc[df["q10"] == "United Kingdomk", "q10"] = "UK"
df.loc[df["q10"] == "New Zealand", "q10"] = "NZ"
df.loc[df["q10"] == " New Zealand", "q10"] = "NZ"
df.loc[df["q10"] == "England, Gb", "q10"] = "UK"
df.loc[df["q10"] == "Canada", "q10"] = "CAN"
df.loc[df["q10"] == "Australia", "q10"] = "AUS"

In [71]:
df.q10.value_counts()

q10
USA                22551
CAN                 1572
UK                  1126
AUS                  318
Germany              174
                   ...  
Bangladesh             1
Eritrea                1
IS                     1
Cuba                   1
Česká republika        1
Name: count, Length: 364, dtype: int64

Minimal cleanup on country has been done.  

In [73]:
df.q1.value_counts()

q1
25-34         12668
35-44          9908
45-54          3193
18-24          1236
55-64           994
65 or over       95
under 18         14
Name: count, dtype: int64

<div class="alert alert-block alert-info">

#### Reflection
Write a short reflection (400-500 words) answering the following: 
- What were the biggest issues you encountered in the messy dataset?
- How did cleaning the dataset improve its usability for machine learning?
- What would happen if we trained a model on the messy dataset vs. the cleaned one?
- Do you feel you skewed or biased the dataset while cleaning it?

[Add you answer to this markdown cell]

One of the biggest issues with a dirty dataset is getting an accurate idea of what the counts are for each category.  For example, with the countries--I only did a base level cleanup on those and I still could have done a lot more.  The large number of nulls also make it difficult to assess the data for value.  Dirty data prevents the ML from being able to learn, as well.  Null isn't a value--it just doesn't exist.  The ML can't do anything with it.  The initial data set appeared to have at least one null in each record.  Moreover, various features contained a mixed bag of values--several different values for USA, including misspellings and added punctuation.  Some of the entries had extra spaces in the free form text entries.  This creates a skewed count and will create an erroneous result from the ML, because it will see each one as a distinct entity.  
<br>
Removing the nulls and removing the columns that do not add value to the dataset, will improve the ML's ability to learn from the data. Without the nulls, the ML will be able to use all the columns to perform training and analytics.  However, given that the dataset shows a bias for whites, women, master's level education or higher, and for those in the US--this indicates that prior to training this dataset, the data person will need to add 'mock' data for men, for other races, other education levels and other nations based on the existing data.  If this is not done, the outcome of the ML will be bias and it will remain that way until the data is balanced.  

Anytime data is touched/updated/cleaned it does skew it.  Had the original dataset been used, much of it would have been useless for ML.  Cleaned, the ML now has more to work with.  The question is whether it skews it for better or worse.  The best thing to do when cleaning data is to make as little impact as possible, by imputing accurate values.  The drawback with this is that it can be excessively time consuming to impute every single questionable dirty value.  


## Deliverables
Upload your Jupyter Notebook to your GitHub repo and then provide a link to that repo in Worlclass. 

References used:
Previous assignments from MSDS 600, 640 and 650